### PR TITLE
[Incremental Imports] Add framework and new tests

### PR DESCRIPTION
**Pull request by davidungar**
Adds a new test framework, and uses it for two multi-module, incremental import tests.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test

File: Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+// A framework for easily adding multi-module tests of the incremental imports
+// It uses enumerations for Modules and SourceFiles because the tests must be
+// able to name particular modules and files, and enum cases give static checkability.
+
+// MARK: - IITF_TestProtocol
+
+/// Each test creates a struct, conforming to this protocol, requiring only the line:
+/// `fileprivate let files = Files<SomeModuleType>`
+/// where `SomeModuleType` is an `enum` conforming to `IITF_ModuleProtocol`.
+ protocol IITF_TestProtocol {
```
Comment (CodaFi) [inline Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift:27]: Swift code generally doesn't use prefixes. You can make a separate test-only module if you are worried about encapsulation.

File: Tests/SwiftDriverTests/IncrementalImportTests.swift [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+/// Ensure that only the users get rebuilt
+fileprivate struct RenameMemberOfImportedStruct: IITF_TestProtocol {
+  fileprivate let files = IncrementalImportTestFramework.Files<Module>()
+}
+
+fileprivate extension RenameMemberOfImportedStruct {
+  enum Module: String, IITF_ModuleProtocol {
+    typealias SourceFile = RenameMemberOfImportedStruct.SourceFile
+    // Must be in build order
+    case ImportedModule, MainModule
```
Comment (CodaFi) [inline Tests/SwiftDriverTests/IncrementalImportTests.swift:41]: Swift 3+ uses lower camel case enum identifiers.

File: Tests/SwiftDriverTests/IncrementalImportTests.swift [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+// MARK: - TransitiveExternals
+
+fileprivate struct HideAndShowFuncInStructAndExtension: IITF_TestProtocol {
+  fileprivate let files = IncrementalImportTestFramework.Files<Module>()
+}
+
+extension HideAndShowFuncInStructAndExtension {
+  fileprivate enum Module: String, IITF_ModuleProtocol {
+    // Must be in build order
+    case ImportedModule, MainModule
```
Comment (CodaFi) [inline Tests/SwiftDriverTests/IncrementalImportTests.swift:112]: Casing

File: Tests/SwiftDriverTests/IncrementalImportTests.swift [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                          """)
+      case .yetAnotherFile: return .stable("""
+                              import \(Module.ImportedModule.name)
+                              func late() {
+                                S()
+                              }
+                              """)
+      case .importedFile:
+        return .mutable([
+          .init("""
```
Comment (CodaFi) [inline Tests/SwiftDriverTests/IncrementalImportTests.swift:175]: This DSL spelling feels like the pendulum has swung really really far in favor of terseness. I would prefer to see a more instantly-readable set of declarations here. For example, type names can be used to signal build steps, labels would help to distinguish the option set.

File: Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    allCases.map {$0.description.versionCount} .reduce(1, *) + 1
+  }
+}
+
+// MARK: - IITF_Namespace
+/// A namespace for all the helpers
+struct IncrementalImportTestFramework {
+}
+// MARK: - SourceFileDescription
+extension IncrementalImportTestFramework {
```
Comment (CodaFi) [inline Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift:206]: Looks like it's more than helpers here if this is contributing to the type of a protocol requirement.

File: Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    Module.allCases.flatMap{ $0.build(files: files, enableIncrementalImports: enableIncrementalImports) }
+  }
+
+  /// Performs a mutation of the mutable source file, and returns the compilations to expect after a subsequent rebuild.
+  /// \c iteration should range from 1 to `iterationCount`, inclusive.
+  private func mutate(enableIncrementalImports: Bool, iteration: Int) -> [Module.SourceFile] {
+    let mutableSourceFiles = Module.SourceFile.allCases.compactMap {
+      $0.description.isMutable ? $0 : nil
+    }
+    assert(mutableSourceFiles.count == 1, "At present, cannot handle more or less than one mutable source")
```
Comment (CodaFi) [inline Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift:86]: Perhaps this suggests that we need a cleaner separation between phase-dependent inputs and stable inputs than the enum. Perhaps there could even be a separate requirement.

**Review by CodaFi — CHANGES_REQUESTED** [review]
I think this framework has the right general structure of a test we'd like here. I do appreciate the explicit phasing here, for example, being a strong part of the interface of the DSL. That said, it is quite terse in certain places (specifying mutations) and quite verbose in others (the number of requirements the client needs to satisfy). It feels like there is a design waiting in the wings here that better utilizes types to bring things those things back into equilibrium. 

tl;dr I think the infrastructure behind the test runner looks good. I think the interface presented to clients could use some work.

**Comment by davidungar** [comment]
@swift-ci please test macos platform

File: Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift (hunk 1 above)
Comment (davidungar) [reply Tests/SwiftDriverTests/Helpers/IncrementalImportTestFramework.swift:27]: Good idea! Will do.

File: Tests/SwiftDriverTests/IncrementalImportTests.swift (hunk 2 above)
Comment (davidungar) [reply Tests/SwiftDriverTests/IncrementalImportTests.swift:41]: Done!

File: Tests/SwiftDriverTests/IncrementalImportTests.swift (hunk 3 above)
Comment (davidungar) [reply Tests/SwiftDriverTests/IncrementalImportTests.swift:112]: Done.

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by davidungar** [comment]
Closing to post a better framework incorporating your feedback
